packages/block-library/src/html/edit.js

```diff
@@ -17,18 +23,45 @@ import Preview from './preview';
 
 export default function HTMLEdit( { attributes, setAttributes, isSelected } ) {
 	const [ isPreview, setIsPreview ] = useState();
+	const [ shouldFocus, setShouldFocus ] = useState( false );
 	const isDisabled = useContext( Disabled.Context );
 
+	const htmlTextareaRef = useRef();
+	const blockRef = useRef();
+
+	const instanceId = useInstanceId( HTMLEdit );
+	const instanceIdDesc = sprintf( 'html-edit-%d-desc', instanceId );
+
 	function switchToPreview() {
 		setIsPreview( true );
+		setShouldFocus( true );
 	}
 
 	function switchToHTML() {
 		setIsPreview( false );
+		setShouldFocus( true );
 	}
 
+	// Effect for managing focus.
+	useEffect( () => {
+		if ( ! shouldFocus ) {
+			return;
+		}
+		if ( isPreview ) {
+			blockRef?.current?.focus();
+		} else {
+			htmlTextareaRef?.current?.focus();
+		}
+	}, [ shouldFocus, isPreview ] );
```

Comment:
Suggestion: Have you tried triggering the focus from `switchToPreview` and `switchToHTML` callbacks? Since these actions trigger DOM change, triggering the focus change also makes sense to me.

While there's nothing wrong with `useEffect`, direct callbacks should have preference.

P.S. I've not tested this approach.

Reply:
@Mamaduka I tried but the ref was unavailable at that stage even calling it after the state change. That's why I opted to track in state.

Reply:
Right, the `htmlTextareaRef ` won't be available when a state update is dispatched.

Reply:
@Mamaduka I think I might try to fix this in writing flow instead. If other blocks render new content, it would be better that writing flow be able to handle it where right now focus handling fails and we have to handle it secondary in blocks.

I'll split my attempt into another PR and just test. It might go somewhere, might not.

Reply:
Makes sense. Ping me if I can help with review or testing.

Reply:
Working in: #54443.